docs/docker_run.rst

```diff
@@ -0,0 +1,190 @@
+automatically added to the list of volumes to mount.
+
+Downloading files from Girder
+-----------------------------
+
+Accessing files held in girder from within a container is straight forward using
+the ``GirderFileIdToVolume`` utility class. One simply provides the file id as
+a argument to the constructor and pass the instance as a container argument.
+
+.. code-block:: python
+
+    docker_run.delay('my/image', pull_image=True,
+        container_args=[GirderFileIdToVolume(file_id)])
+
+The ``GirderFileIdToVolume`` instance will take care of downloading the file from
+Girder and passes the path it was download to the docker containers entrypoint as an argument.
+
+If no ``volume`` parameter is specified then the file will be downloading to the
+task temporary volume. The file can also be downloaded to a specific ``Volume`` by
+specifying a volume parameter, as follows:
+
+.. code-block:: python
+
+    vol = Volume(host_path, container_path)
+    docker_run.delay('my/image', pull_image=True,
+        container_args=[GirderFileIdToVolume(file_id,volume=vol)])
+
+If the being downloaded is particularly large you may want to consider streaming
+it into the container using a named pipe. See `Streaming Girder files into a container`_
+for more details.
+
+Uploading files to Girder items
+-------------------------------
+
+Utility classes are also provide to simplify uploading files generated by a the
+docker container. The ``GirderUploadVolumePathToItem`` provides the functionality
+to upload a file to an item. In the example below, we using the ``VolumePath``
+utility class to define a file path that we then pass to the docker container.
+The docker container can write data to this file path. As well as passing the
+``VolumePath`` instance as a container argument we also pass it to ``GirderUploadVolumePathToItem``
+, the ``GirderUploadVolumePathToItem`` instance is add to ``girder_result_hooks``. This
+tell girder_work to upload the file path to the item id provided when the docker
+container has finished running.
+
+
+.. code-block:: python
+
+    volumepath = VolumePath('write_data_to_be_upoaded.txt')
+    docker_run.delay('my/image', pull_image=True, container_args=[volumepath],
+        girder_result_hooks=[GirderUploadVolumePathToItem(volumepath, item_id)])
+
+
+Using named pipes to stream data in and out of containers
+---------------------------------------------------------
+
+girder_worker uses named pipes as a language agnostic way of streaming data in
+and out of docker containers. Basically a named pipe is created at a path that is
+mounted into the container. This allows the container to open that pipe for read or
+write and similarly the girder_worker infrastructure can open the pipe on the host,
+thus allow data write and read from the container.
+
+The are two utility classes using to represent as named pipe, ``NamedOutputPipe``
+and ``NamedInputPipe``
```

Comment:
Let's link to API docs for these.

Reply:
@zachmullen  How do you link to a particular class in the API docks?